dbms/tests/integration/test_storage_s3/test.py

```diff
@@ -0,0 +1,156 @@
+                cluster.preserving_data_port = data['preserving_data_port']
+                cluster.multipart_preserving_data_port = data['multipart_preserving_data_port']
+                cluster.redirecting_preserving_data_port = data['redirecting_preserving_data_port']
+            except:
+                logging.error(traceback.format_exc())
+                time.sleep(0.5)
+            else:
+                break
+        else:
+            assert False, 'Could not initialize mock server'
+
+        yield cluster
+
+    finally:
+        cluster.shutdown()
+
+
+def run_query(instance, query, stdin=None):
+    logging.info('Running query "{}"...'.format(query))
+    result = instance.query(query, stdin=stdin)
+    logging.info('Query finished')
+    return result
+
+def test_get_with_redirect(started_cluster):
+    instance = started_cluster.instances['dummy']
+    format = 'column1 UInt32, column2 UInt32, column3 UInt32'
+
+    put_communication_data(started_cluster, '=== Get with redirect test ===')
+    query = "select *, column1*column2*column3 from s3('http://{}:{}/', 'CSV', '{}')".format(started_cluster.mock_host, started_cluster.redirecting_to_http_port, format)
+    stdout = run_query(instance, query)
+    assert list(map(str.split, stdout.splitlines())) == [
+        ['42', '87', '44', '160776'],
+        ['55', '33', '81', '147015'],
+        ['1', '0', '9', '0'],
+    ]
+    
+def test_put(started_cluster):
+    instance = started_cluster.instances['dummy']
+    format = 'column1 UInt32, column2 UInt32, column3 UInt32'
+
+    logging.info('Phase 3')
+    put_communication_data(started_cluster, '=== Put test ===')
+    values = '(1, 2, 3), (3, 2, 1), (78, 43, 45)'
+    put_query = "insert into table function s3('http://{}:{}/{}/test.csv', 'CSV', '{}') values {}".format(started_cluster.mock_host, started_cluster.preserving_data_port, started_cluster.bucket, format, values)
+    run_query(instance, put_query)
+    data = get_communication_data(started_cluster)
+    received_data_completed = data['received_data_completed']
+    received_data = data['received_data']
+    finalize_data = data['finalize_data']
+    finalize_data_query = data['finalize_data_query']
+    assert received_data[-1].decode() == '1,2,3\n3,2,1\n78,43,45\n'
+    assert received_data_completed
+    assert finalize_data == '<CompleteMultipartUpload><Part><PartNumber>1</PartNumber><ETag>hello-etag</ETag></Part></CompleteMultipartUpload>'
+    assert finalize_data_query == 'uploadId=TEST'
+    
+def test_put_csv(started_cluster):
+    instance = started_cluster.instances['dummy']
+    format = 'column1 UInt32, column2 UInt32, column3 UInt32'
+
+    put_communication_data(started_cluster, '=== Put test CSV ===')
+    put_query = "insert into table function s3('http://{}:{}/{}/test.csv', 'CSV', '{}') format CSV".format(started_cluster.mock_host, started_cluster.preserving_data_port, started_cluster.bucket, format)
+    csv_data = '8,9,16\n11,18,13\n22,14,2\n'
```

Comment:
Could we have one type of quotes for strings in file?

Reply:
64d49e4565